pkg/libovsdb/schema.go

```diff
@@ -71,3 +79,344 @@ func (schema DatabaseSchema) validateOperations(operations ...Operation) bool {
+	if err := json.Unmarshal(colJSON.TypeRawMsg, &typeString); err == nil {
+		if !isAtomicType(typeString) {
+			return fmt.Errorf("schema contains unknown atomic type %s", typeString)
+		}
+		// This was an easy one. Use the string as our 'extended' type
+		column.Type = typeString
+		return nil
+	}
+
+	// 'type' can be an object defined as:
+	// "key": <base-type>                 required
+	// "value": <base-type>               optional
+	// "min": <integer>                   optional (default: 1)
+	// "max": <integer> or "unlimited"    optional (default: 1)
+	column.TypeObj = &ColumnType{
+		Key:   &BaseType{},
+		Value: nil,
+		Max:   1,
+		Min:   1,
+	}
+
+	// ColumnTypeJSON is used to dynamically decode the ColumnType
+	type ColumnTypeJSON struct {
+		KeyRawMsg   *json.RawMessage `json:"key,omitempty"`
+		ValueRawMsg *json.RawMessage `json:"value,omitempty"`
+		Min         int              `json:"min,omitempty"`
+		MaxRawMsg   *json.RawMessage `json:"max,omitempty"`
+	}
+	colTypeJSON := ColumnTypeJSON{
+		Min: 1,
+	}
+
+	if err := json.Unmarshal(colJSON.TypeRawMsg, &colTypeJSON); err != nil {
+		return fmt.Errorf("cannot parse type object: %s", err)
+	}
+
+	// Now we have to unmarshall some fields manually because they can store
+	// values of different types. Also, in order to really know what native
+	// type can store a value of this column, the RFC defines some logic based
+	// on the values of 'type'. So, in addition to manually unmarshalling, let's
+	// figure out what is the real native type and store it in column.Type for
+	// ease of use.
+
+	// 'max' can be an integer or the string "unlimmited". To simplify, use -1
+	// as unlimited
+	if colTypeJSON.MaxRawMsg != nil {
+		var maxString string
+		if err := json.Unmarshal(*colTypeJSON.MaxRawMsg, &maxString); err == nil {
+			if maxString == "unlimited" {
+				column.TypeObj.Max = Unlimited
+			} else {
+				return fmt.Errorf("unknown max value %s", maxString)
+			}
+		} else if err := json.Unmarshal(*colTypeJSON.MaxRawMsg, &column.TypeObj.Max); err != nil {
+			return fmt.Errorf("cannot parse max field: %s", err)
+		}
+	}
+	column.TypeObj.Min = colTypeJSON.Min
+
+	// 'key' and 'value' can, themselves, be a string or a BaseType.
+	// key='<atomic_type>' is equivalent to 'key': {'type': '<atomic_type>'}
+	// To simplify things a bit, we'll translate the former to the latter
+
+	if err := json.Unmarshal(*colTypeJSON.KeyRawMsg, &column.TypeObj.Key.Type); err != nil {
+		if err := json.Unmarshal(*colTypeJSON.KeyRawMsg, column.TypeObj.Key); err != nil {
+			return fmt.Errorf("cannot parse key object: %s", err)
+		}
+		if err := column.TypeObj.Key.parseEnum(*colTypeJSON.KeyRawMsg); err != nil {
+			return err
+		}
+	}
+
+	if !isAtomicType(column.TypeObj.Key.Type) {
+		return fmt.Errorf("schema contains unknown atomic type %s", column.TypeObj.Key.Type)
+	}
+
+	// 'value' is optional. If it exists, we know the real native type is a map
+	if colTypeJSON.ValueRawMsg != nil {
```

Comment:
if it can simplify, the map's key are strings

Reply:
libovsdb code